packages/core/src/ui/UICanvas.ts

```diff
@@ -0,0 +1,495 @@
+import { Component } from "../Component";
+import { DependentMode, dependentComponents } from "../ComponentsDependencies";
+import { Entity, EntityModifyFlags } from "../Entity";
+import { RenderContext } from "../RenderPipeline/RenderContext";
+import { RenderElement } from "../RenderPipeline/RenderElement";
+import { assignmentClone, ignoreClone } from "../clone/CloneManager";
+import { ComponentType } from "../enums/ComponentType";
+import { HitResult } from "../physics";
+import { CanvasGroup } from "./CanvasGroup";
+import { UIRenderer } from "./UIRenderer";
+import { UITransform } from "./UITransform";
+import { CanvasRenderMode } from "./enums/CanvasRenderMode";
+import { ResolutionAdaptationStrategy } from "./enums/ResolutionAdaptationStrategy";
+
+@dependentComponents(UITransform, DependentMode.AutoAdd)
+export class UICanvas extends Component {
+  /** @internal */
+  @ignoreClone
+  _canvasIndex: number = -1;
+  /** @internal */
+  @ignoreClone
+  _isRootCanvas: boolean = false;
+  /** @internal */
+  @ignoreClone
+  _renderElement: RenderElement;
+  /** @internal */
+  @ignoreClone
+  _sortDistance: number = 0;
+
+  @assignmentClone
+  private _renderMode = CanvasRenderMode.WorldSpace;
+  @assignmentClone
+  private _renderCamera: Camera;
+  @assignmentClone
+  private _resolutionAdaptationStrategy = ResolutionAdaptationStrategy.BothAdaptation;
+  @assignmentClone
+  private _sortOrder: number = 0;
+  @assignmentClone
+  private _distance: number = 10;
+  private _renderers: UIRenderer[] = [];
+  private _transform: UITransform;
+  private _referenceResolution: Vector2 = new Vector2(800, 600);
+  private _enableBlocked: boolean = true;
+  private _parents: Entity[] = [];
+
+  /** @internal */
+  get renderers(): UIRenderer[] {
+    const renderers = this._renderers;
+    renderers.length = 0;
+    this._walk(this.entity, renderers);
+    return renderers;
+  }
+
+  get enableBlocked(): boolean {
+    return this._enableBlocked;
+  }
+
+  set enableBlocked(value: boolean) {
+    this._enableBlocked = value;
+  }
+
+  get referenceResolution(): Vector2 {
+    return this._referenceResolution;
+  }
+
+  set referenceResolution(val: Vector2) {
+    const { _referenceResolution: referenceResolution } = this;
+    if (referenceResolution === val) return;
+    (referenceResolution.x !== val.x || referenceResolution.y !== val.y) && referenceResolution.copyFrom(val);
+  }
+
+  get renderMode(): CanvasRenderMode {
+    return this._renderMode;
+  }
+
+  set renderMode(mode: CanvasRenderMode) {
+    let preMode = this._renderMode;
+    if (preMode !== mode) {
+      this._renderMode = mode;
+      if (this._isRootCanvas) {
+        const camera = this._renderCamera;
+        preMode =
+          preMode === CanvasRenderMode.ScreenSpaceCamera && !camera ? CanvasRenderMode.ScreenSpaceOverlay : preMode;
+        mode = mode === CanvasRenderMode.ScreenSpaceCamera && !camera ? CanvasRenderMode.ScreenSpaceOverlay : mode;
```

Comment:
The logic itself is relatively simple, and there are only two calls here. Adding function call overhead for this may not be a very cost-effective thing.